### DISCUSSION TITLE
Lesson 12: Invariant Tests Pass on Revert

### DISCUSSION THREAD

**Discussion by AztecWarlord** [Cyfrin/foundry-full-course-cu#827] [General]
Invariant tests are giving a passing result when only one run is successful and a call is reverted. Noting that my ```.toml``` ```fail_on_revert = true```.
Here is result:
```Running 1 test for test/fuzz/Invariants.t.sol:InvariantsTest
[PASS] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 1, calls: 2, reverts: 1). ```

**Comment by 00decree** [comment]
i'm getting the same result.

```
forge test --mt invariant_protocolMustHaveMoreValueThanTotalSupply
[⠆] Compiling...
[⠃] Compiling 2 files with 0.8.19
[⠊] Solc 0.8.19 finished in 1.63s
Compiler run successful!

Running 1 test for test/fuzz/InvariantsTest.t.sol:InvariantTest
[PASS] invariant_protocolMustHaveMoreValueThanTotalSupply() (runs: 1, calls: 2, reverts: 1)
Test result: ok. 1 passed; 0 failed; 0 skipped; finished in 6.76ms

Ran 1 test suites: 1 tests passed, 0 failed, 0 skipped (1 total tests)
```

Probably gotta do with this issue reported on Foundry here 
https://github.com/foundry-rs/foundry/issues/5726

**Reply by Packagefather** [reply]
check this out. Patrick did mention that we consider any test with revert in the data not equal to 0 as failed for now, until the bug is fixed. https://github.com/foundry-rs/foundry/issues/5726#issuecomment-1732399838

**Comment by theirrationalone** [comment]
It'll not give any sequence error reverts until invariant asserts persist to

`wethTokenBalance + wbtcTokenBalance >= totalsupply`


Since handler has no any syntactic, run time, error, it will always get passed. Therefore you might have reverts mitigations implemented into handler functions i.e., bound, MustBeMoreThanZero -> if (isAmountZero == true) return;, etc. 


Have you noticed... 

`MustBeMoreThanZero -> if (isAmountZero == true) return;`

Has reverted and also returned for some random data into handler functions.

Now, Let's assume, Unfortunately we got a revert sequence. As we know, Revert throws an error and therefore dismisses the execution. So execution might got dismissed. Due to the flow, execution might came to invariant_test from handler, after dismissing handler functions execution. And retrieved & asserted balances as well as total supply. No matter handler got reverted, if accumulated total balances is greater than Or equals to total supply, Foundry could not consider that a sequence error. 

Things to keep in mind:

1. A valid sequence catch is what, when intended assertions into invariant get blown. 

2. If handler has any error syntax/runtime/revert with no bound Or specific mitigations. 


You can test these things out by implementing for-sure breach intentionally. Or you can follow Patrick again. 

If you're sure for these things above, and still getting reverts but foundry isn't catching sequences. You can raise an issue at foundry's github Or can ask/call Patrick via twitter Or tag him here. 

Sorry i'm wrong ❌ here ;`) 😅
